src/drawers/canvasDrawer.ts

```diff
@@ -20,6 +20,9 @@ export type CanvasDrawStep = (
  * This class is immutable (but has internal state) and shouldn't be extended.
  */
 export class CanvasDrawer implements IDrawer {
+  protected _lastDrawnData: any[];
+  protected _lastProjector: AttributeToAppliedProjector;
```

Comment:
The reason i did it this way was to avoid too much unnecessary duplicate computation.

The SVG drawer doesn't need to invoke `resolveAttributesSubsetWithStyles` here -- it already does it when it renders and it'll just use the element's BBox. Furthermore, the various canvas drawers user `resolveAttributesSubsetWithStyles` in different ways or not at all. So, this approach allows us to only compute the bounds lazily. And even though we are storing this state, it's a shallow copy--whereas precomputing with `resolveAttributesSubsetWithStyles` has a perf penalty.

I agree that there are several improvements we can make across the various places that use entity bounds. I agree we should consolidate. I think both drawers (svg & canvas) as well as the plot code that iterates over entities could benefit from a refactor here.

This change was made without refactoring mainly for expediency.